File: src/commands.js
const { SlashCommandBuilder, PermissionFlagsBits, EmbedBuilder } = require('discord.js');

// ── Embed Helpers ─────────────────────────────────────────────────────────────
const ok   = (title, desc = '') => new EmbedBuilder().setTitle(`✅ ${title}`).setDescription(desc).setColor(0x2ECC71);
const info = (title, desc = '') => new EmbedBuilder().setTitle(`ℹ️ ${title}`).setDescription(desc).setColor(0x3498DB);
const err  = (title, desc = '') => new EmbedBuilder().setTitle(`❌ ${title}`).setDescription(desc).setColor(0xE74C3C);

// ── Command Definitions ───────────────────────────────────────────────────────
const commands = [
  // /automod status
  new SlashCommandBuilder()
    .setName('automod')
    .setDescription('🛡️ Kelola pengaturan AutoMod')
    .setDefaultMemberPermissions(PermissionFlagsBits.Administrator)
    .addSubcommand(s => s.setName('status').setDescription('Tampilkan konfigurasi AutoMod saat ini'))
    .addSubcommand(s => s.setName('help').setDescription('Tampilkan semua perintah AutoMod'))
    .addSubcommand(s => s
      .setName('log')
      .setDescription('Riwayat penghapusan otomatis')
      .addIntegerOption(o => o.setName('limit').setDescription('Jumlah entri (1–20)').setMinValue(1).setMaxValue(20))
    )
    // server subcommands
    .addSubcommandGroup(g => g
      .setName('server')
      .setDescription('Konfigurasi server-wide')
      .addSubcommand(s => s
        .setName('links')
        .setDescription('Toggle filter link server-wide')
        .addBooleanOption(o => o.setName('enabled').setDescription('Aktifkan atau matikan').setRequired(true))
      )
      .addSubcommand(s => s
        .setName('images')
        .setDescription('Toggle filter gambar server-wide')
        .addBooleanOption(o => o.setName('enabled').setDescription('Aktifkan atau matikan').setRequired(true))
      )
      .addSubcommand(s => s
        .setName('logchannel')
        .setDescription('Set channel untuk mod-log')
        .addChannelOption(o => o.setName('channel').setDescription('Channel tujuan log').setRequired(true))
      )
      .addSubcommand(s => s
        .setName('warn')
        .setDescription('Toggle DM peringatan ke user saat pesan dihapus')
        .addBooleanOption(o => o.setName('enabled').setDescription('Aktifkan atau matikan').setRequired(true))
      )
    )
    // channel subcommands
    .addSubcommandGroup(g => g
      .setName('channel')
      .setDescription('Konfigurasi per channel')
      .addSubcommand(s => s
        .setName('links')
        .setDescription('Toggle filter link di channel tertentu')
        .addChannelOption(o => o.setName('channel').setDescription('Target channel').setRequired(true))
        .addBooleanOption(o => o.setName('enabled').setDescription('Aktifkan atau matikan').setRequired(true))
      )
      .addSubcommand(s => s
        .setName('images')
        .setDescription('Toggle filter gambar di channel tertentu')
        .addChannelOption(o => o.setName('channel').setDescription('Target channel').setRequired(true))
        .addBooleanOption(o => o.setName('enabled').setDescription('Aktifkan atau matikan').setRequired(true))
      )
      .addSubcommand(s => s
        .setName('whitelist_add')
        .setDescription('Izinkan domain tertentu di channel')
        .addChannelOption(o => o.setName('channel').setDescription('Target channel').setRequired(true))
        .addStringOption(o => o.setName('domain').setDescription('Contoh: youtube.com').setRequired(true))
      )
      .addSubcommand(s => s
        .setName('whitelist_remove')
        .setDescription('Hapus domain dari whitelist channel')
        .addChannelOption(o => o.setName('channel').setDescription('Target channel').setRequired(true))
        .addStringOption(o => o.setName('domain').setDescription('Domain yang ingin dihapus').setRequired(true))
      )
    )
    // role subcommands
    .addSubcommandGroup(g => g
      .setName('role')
      .setDescription('Kelola role yang bypass semua filter')
      .addSubcommand(s => s
        .setName('add')
        .setDescription('Tambah role ke daftar exempt')
        .addRoleOption(o => o.setName('role').setDescription('Role yang akan dibebaskan').setRequired(true))
      )
      .addSubcommand(s => s
        .setName('remove')
        .setDescription('Hapus role dari daftar exempt')
        .addRoleOption(o => o.setName('role').setDescription('Role yang ingin dihapus').setRequired(true))
      )
      .addSubcommand(s => s.setName('list').setDescription('Lihat semua role yang exempt'))
    )
].map(c => c.toJSON());

// ── Command Handlers ──────────────────────────────────────────────────────────
async function handleCommand(interaction, db) {
  if (!interaction.isChatInputCommand()) return;
  if (interaction.commandName !== 'automod') return;

  const sub   = interaction.options.getSubcommand(false);
  const group = interaction.options.getSubcommandGroup(false);
  const guild = interaction.guild;

  try {
    // ── Top-level subcommands ──────────────────────────────────────────────
    if (!group) {
      if (sub === 'status') return await handleStatus(interaction, db);
      if (sub === 'help')   return await handleHelp(interaction);
      if (sub === 'log')    return await handleLog(interaction, db);
    }

    // ── server group ───────────────────────────────────────────────────────
    if (group === 'server') {
      const enabled = interaction.options.getBoolean('enabled');
      const onOff   = enabled ? '🟢 aktif' : '🔴 nonaktif';

      if (sub === 'links') {
        db.upsertGuild(guild.id, { delete_links: enabled ? 1 : 0 });
        return interaction.reply({ embeds: [ok('Server Link Filter', `Filter link **${onOff}** server-wide`)], ephemeral: true });
      }
      if (sub === 'images') {
        db.upsertGuild(guild.id, { delete_images: enabled ? 1 : 0 });
        return interaction.reply({ embeds: [ok('Server Image Filter', `Filter gambar **${onOff}** server-wide`)], ephemeral: true });
      }
      if (sub === 'logchannel') {
        const ch = interaction.options.getChannel('channel');
        db.upsertGuild(guild.id, { log_channel: ch.id });
        return interaction.reply({ embeds: [ok('Log Channel Set', `Mod log dikirim ke ${ch}`)], ephemeral: true });
      }
      if (sub === 'warn') {
        db.upsertGuild(guild.id, { warn_user: enabled ? 1 : 0 });
        return interaction.reply({ embeds: [ok('User Warning', `DM peringatan **${onOff}**`)], ephemeral: true });
      }
    }

    // ── channel group ──────────────────────────────────────────────────────
    if (group === 'channel') {
      const ch      = interaction.options.getChannel('channel');
      const enabled = interaction.options.getBoolean('enabled');
      const onOff   = enabled ? '🟢 aktif' : '🔴 nonaktif';

      if (sub === 'links') {
        db.upsertChannel(ch.id, guild.id, { delete_links: enabled ? 1 : 0 });
        return interaction.reply({ embeds: [ok('Channel Link Filter', `Filter link **${onOff}** di ${ch}`)], ephemeral: true });
      }
      if (sub === 'images') {
        db.upsertChannel(ch.id, guild.id, { delete_images: enabled ? 1 : 0 });
        return interaction.reply({ embeds: [ok('Channel Image Filter', `Filter gambar **${onOff}** di ${ch}`)], ephemeral: true });
      }
      if (sub === 'whitelist_add') {
        let domain = interaction.options.getString('domain').toLowerCase().trim();
        domain = domain.replace(/^https?:\/\//, '').split('/')[0];
        db.addWhitelist(ch.id, guild.id, domain);
        return interaction.reply({ embeds: [ok('Whitelist Updated', `\`${domain}\` ditambahkan ke whitelist ${ch}`)], ephemeral: true });
      }
      if (sub === 'whitelist_remove') {
        const domain = interaction.options.getString('domain').toLowerCase().trim();
        db.removeWhitelist(ch.id, guild.id, domain);
        return interaction.reply({ embeds: [ok('Whitelist Updated', `\`${domain}\` dihapus dari whitelist ${ch}`)], ephemeral: true });
      }
    }

    // ── role group ─────────────────────────────────────────────────────────
    if (group === 'role') {
      if (sub === 'add') {
        const role = interaction.options.getRole('role');
        db.addRole(guild.id, role.id);
        return interaction.reply({ embeds: [ok('Exempt Role Added', `${role} kini bebas dari semua filter`)], ephemeral: true });
      }
      if (sub === 'remove') {
        const role = interaction.options.getRole('role');
        db.removeRole(guild.id, role.id);
        return interaction.reply({ embeds: [ok('Exempt Role Removed', `${role} dihapus dari daftar exempt`)], ephemeral: true });
      }
      if (sub === 'list') {
        const roleIds  = db.getRoles(guild.id);
        const mentions = roleIds.map(id => guild.roles.cache.get(id)).filter(Boolean).map(r => r.toString());
        return interaction.reply({
          embeds: [info('Exempt Roles', mentions.length ? mentions.join(', ') : 'Tidak ada role exempt.')],
          ephemeral: true,
        });
      }
    }

  } catch (e) {
    console.error('[Command Error]', e);
    const payload = { embeds: [err('Error', `Terjadi kesalahan: \`${e.message}\``)], ephemeral: true };
    if (interaction.replied || interaction.deferred) {
      await interaction.followUp(payload).catch(() => {});
    } else {
      await interaction.reply(payload).catch(() => {});
    }
  }
}

// ── /automod status ───────────────────────────────────────────────────────────
async function handleStatus(interaction, db) {
  await interaction.deferReply({ ephemeral: true });
  const guild    = interaction.guild;
  const gcfg     = db.getGuild(guild.id);
  const channels = db.getAllChannels(guild.id);
  const roleIds  = db.getRoles(guild.id);

  const embed = new EmbedBuilder()
    .setTitle('🛡️ AutoMod Status')
    .setColor(0x5865F2)
    .setThumbnail(guild.iconURL());

  const gLinks  = gcfg?.delete_links  ? '🟢 On' : '🔴 Off';
  const gImages = gcfg?.delete_images ? '🟢 On' : '🔴 Off';
  const gWarn   = gcfg?.warn_user     ? '🟢 On' : '🔴 Off';
  const logCh   = gcfg?.log_channel   ? `<#${gcfg.log_channel}>` : 'Belum diset';

  embed.addFields({
    name: '🌐 Server-wide',
    value: `**Filter Link:** ${gLinks}\n**Filter Gambar:** ${gImages}\n**Warn Users:** ${gWarn}\n**Log Channel:** ${logCh}`,
    inline: false,
  });

  const roleMentions = roleIds.map(id => guild.roles.cache.get(id)).filter(Boolean).map(r => r.toString());
  embed.addFields({
    name: '🔓 Exempt Roles',
    value: roleMentions.length ? roleMentions.join(', ') : 'Tidak ada',
    inline: false,
  });

  if (channels.length > 0) {
    const lines = channels.slice(0, 10).map(ch => {
      const flags = [];
      if (ch.delete_links)  flags.push('🔗 links');
      if (ch.delete_images) flags.push('🖼️ images');
      const wl = ch.whitelist ? ch.whitelist.split(',').filter(Boolean) : [];
      if (wl.length) flags.push(`✅ \`${wl.slice(0, 3).join(', ')}\``);
      return flags.length ? `<#${ch.channel_id}>: ${flags.join(' | ')}` : null;
    }).filter(Boolean);

    if (lines.length) {
      embed.addFields({ name: '📋 Channel Overrides', value: lines.join('\n'), inline: false });
    }
  }

  embed.setFooter({ text: `Guild ID: ${guild.id}` });
  await interaction.editReply({ embeds: [embed] });
}

// ── /automod log ──────────────────────────────────────────────────────────────
async function handleLog(interaction, db) {
  await interaction.deferReply({ ephemeral: true });
  const limit = interaction.options.getInteger('limit') || 5;
  const rows  = db.getLog(interaction.guild.id, limit);

  if (!rows.length) {
    return interaction.editReply({ embeds: [info('Delete Log', 'Belum ada entri.')] });
  }

  const embed = new EmbedBuilder().setTitle('📋 Recent Deletions').setColor(0xF39C12);
  for (const row of rows) {
    const member  = interaction.guild.members.cache.get(row.user_id);
    const channel = interaction.guild.channels.cache.get(row.channel_id);
    embed.addFields({
      name: `🕐 ${row.deleted_at} UTC`,
      value: `**User:** ${member ? member.user.tag : `ID:${row.user_id}`}\n**Channel:** ${channel ? `#${channel.name}` : row.channel_id}\n**Reason:** ${row.reason}`,
      inline: false,
    });
  }
  await interaction.editReply({ embeds: [embed] });
}

// ── /automod help ─────────────────────────────────────────────────────────────
async function handleHelp(interaction) {
  const embed = new EmbedBuilder()
    .setTitle('🛡️ AutoMod Bot — Help')
    .setDescription('Auto-delete link dan gambar dengan kontrol penuh.')
    .setColor(0x5865F2)
    .addFields(
      { name: '📊 Umum', value: '`/automod status` — Lihat config\n`/automod log` — Riwayat hapus', inline: false },
      { name: '🌐 Server `/automod server`', value: '`links on/off` — Filter link\n`images on/off` — Filter gambar\n`logchannel #ch` — Set log channel\n`warn on/off` — DM ke user', inline: false },
      { name: '📌 Channel `/automod channel`', value: '`links #ch on/off` — Filter link\n`images #ch on/off` — Filter gambar\n`whitelist_add #ch domain` — Izinkan domain\n`whitelist_remove #ch domain` — Hapus domain', inline: false },
      { name: '🔓 Role `/automod role`', value: '`add @role` — Tambah exempt\n`remove @role` — Hapus exempt\n`list` — Lihat semua', inline: false },
      { name: '💡 Tips', value: '• Admin selalu exempt\n• Config channel override server-wide\n• Bot butuh permission **Manage Messages**', inline: false },
    )
    .setFooter({ text: 'Semua command butuh permission Administrator' });

  await interaction.reply({ embeds: [embed], ephemeral: true });
}

module.exports = { commands, handleCommand };

File: cogs/commands.py
import discord
from discord import app_commands
from discord.ext import commands
import logging

log = logging.getLogger("AutoMod.Commands")


# ── Embed Helpers ─────────────────────────────────────────────────────────────
def ok(title: str, desc: str = "") -> discord.Embed:
    return discord.Embed(title=f"✅ {title}", description=desc, color=0x2ECC71)

def info(title: str, desc: str = "") -> discord.Embed:
    return discord.Embed(title=f"ℹ️ {title}", description=desc, color=0x3498DB)

def err(title: str, desc: str = "") -> discord.Embed:
    return discord.Embed(title=f"❌ {title}", description=desc, color=0xE74C3C)


# ── Cog ───────────────────────────────────────────────────────────────────────
class Commands(commands.Cog, name="Commands"):

    def __init__(self, bot):
        self.bot = bot
        self.db = bot.db

    automod = app_commands.Group(
        name="automod",
        description="🛡️ Manage AutoMod settings",
        default_member_permissions=discord.Permissions(administrator=True),
    )

    # ── /automod status ───────────────────────────────────────────────────────
    @automod.command(name="status", description="Tampilkan konfigurasi AutoMod saat ini")
    async def status(self, interaction: discord.Interaction):
        await interaction.response.defer(ephemeral=True)
        guild = interaction.guild
        gcfg  = await self.db.get_guild(guild.id)
        channels = await self.db.get_all_channels(guild.id)
        roles    = await self.db.get_roles(guild.id)

        embed = discord.Embed(title="🛡️ AutoMod Status", color=0x5865F2)
        embed.set_thumbnail(url=guild.icon.url if guild.icon else None)

        g_links  = bool(gcfg["delete_links"])  if gcfg else False
        g_images = bool(gcfg["delete_images"]) if gcfg else False
        g_warn   = bool(gcfg["warn_user"])      if gcfg else False
        g_log    = (
            guild.get_channel(gcfg["log_channel"]).mention
            if gcfg and gcfg["log_channel"] else "Not set"
        )

        embed.add_field(
            name="🌐 Server-wide",
            value=(
                f"**Delete Links:** {'🟢 On' if g_links else '🔴 Off'}\n"
                f"**Delete Images:** {'🟢 On' if g_images else '🔴 Off'}\n"
                f"**Warn Users:** {'🟢 On' if g_warn else '🔴 Off'}\n"
                f"**Log Channel:** {g_log}"
            ),
            inline=False,
        )

        role_mentions = [
            guild.get_role(rid).mention
            for rid in roles if guild.get_role(rid)
        ]
        embed.add_field(
            name="🔓 Exempt Roles",
            value=", ".join(role_mentions) if role_mentions else "None",
            inline=False,
        )

        if channels:
            ch_lines = []
            for ch in channels[:10]:
                channel = guild.get_channel(ch["channel_id"])
                if not channel:
                    continue
                flags = []
                if ch["delete_links"]:  flags.append("🔗 links")
                if ch["delete_images"]: flags.append("🖼️ images")
                wl = [d for d in ch["whitelist"].split(",") if d] if ch["whitelist"] else []
                if wl: flags.append(f"✅ `{', '.join(wl[:3])}`")
                if flags:
                    ch_lines.append(f"{channel.mention}: {' | '.join(flags)}")
            if ch_lines:
                embed.add_field(
                    name="📋 Channel Overrides",
                    value="\n".join(ch_lines),
                    inline=False,
                )

        embed.set_footer(text=f"Guild ID: {guild.id}")
        await interaction.followup.send(embed=embed, ephemeral=True)

    # ── /automod help ─────────────────────────────────────────────────────────
    @automod.command(name="help", description="Tampilkan semua perintah AutoMod")
    async def help_cmd(self, interaction: discord.Interaction):
        embed = discord.Embed(
            title="🛡️ AutoMod Bot — Help",
            description="Auto-delete links dan gambar dengan kontrol penuh.",
            color=0x5865F2,
        )
        embed.add_field(
            name="📊 Umum",
            value="`/automod status` — Lihat config\n`/automod log [limit]` — Riwayat hapus",
            inline=False,
        )
        embed.add_field(
            name="🌐 Server-wide `/automod server`",
            value=(
                "`links on/off` — Filter link seluruh server\n"
                "`images on/off` — Filter gambar seluruh server\n"
                "`logchannel #ch` — Set channel mod-log\n"
                "`warn on/off` — DM peringatan ke user"
            ),
            inline=False,
        )
        embed.add_field(
            name="📌 Per Channel `/automod channel`",
            value=(
                "`links #ch on/off` — Filter link di channel\n"
                "`images #ch on/off` — Filter gambar di channel\n"
                "`whitelist_add #ch domain` — Izinkan domain\n"
                "`whitelist_remove #ch domain` — Hapus domain"
            ),
            inline=False,
        )
        embed.add_field(
            name="🔓 Exempt Role `/automod role`",
            value=(
                "`add @role` — Tambah role exempt\n"
                "`remove @role` — Hapus role exempt\n"
                "`list` — Lihat semua role exempt"
            ),
            inline=False,
        )
        embed.add_field(
            name="💡 Tips",
            value=(
                "• Admin selalu exempt\n"
                "• Config channel override server-wide\n"
                "• Bot butuh permission **Manage Messages**"
            ),
            inline=False,
        )
        embed.set_footer(text="Semua command butuh permission Administrator")
        await interaction.response.send_message(embed=embed, ephemeral=True)

    # ── /automod log ──────────────────────────────────────────────────────────
    @automod.command(name="log", description="Riwayat penghapusan otomatis")
    @app_commands.describe(limit="Jumlah entri (1–20, default 5)")
    async def show_log(
        self,
        interaction: discord.Interaction,
        limit: app_commands.Range[int, 1, 20] = 5,
    ):
        await interaction.response.defer(ephemeral=True)
        rows = await self.db.get_log(interaction.guild.id, limit)

        if not rows:
            await interaction.followup.send(
                embed=info("Delete Log", "Belum ada entri."), ephemeral=True
            )
            return

        embed = discord.Embed(title="📋 Recent Deletions", color=0xF39C12)
        for row in rows:
            user    = interaction.guild.get_member(row["user_id"])
            channel = interaction.guild.get_channel(row["channel_id"])
            embed.add_field(
                name=f"🕐 {row['deleted_at']} UTC",
                value=(
                    f"**User:** {str(user) if user else f'ID:{row[\"user_id\"]}'}\n"
                    f"**Channel:** #{channel.name if channel else row['channel_id']}\n"
                    f"**Reason:** {row['reason']}"
                ),
                inline=False,
            )
        await interaction.followup.send(embed=embed, ephemeral=True)

    # ────────────────────────────────────────────────────────────────────────
    # SERVER GROUP
    # ────────────────────────────────────────────────────────────────────────
    server_group = app_commands.Group(
        name="server", description="Konfigurasi AutoMod server-wide", parent=automod
    )

    @server_group.command(name="links", description="Toggle filter link server-wide")
    @app_commands.describe(enabled="Aktifkan atau matikan")
    async def server_links(self, interaction: discord.Interaction, enabled: bool):
        await self.db.upsert_guild(interaction.guild.id, delete_links=int(enabled))
        await interaction.response.send_message(
            embed=ok("Server Link Filter", f"Filter link **{'aktif' if enabled else 'nonaktif'}** server-wide"),
            ephemeral=True,
        )

    @server_group.command(name="images", description="Toggle filter gambar server-wide")
    @app_commands.describe(enabled="Aktifkan atau matikan")
    async def server_images(self, interaction: discord.Interaction, enabled: bool):
        await self.db.upsert_guild(interaction.guild.id, delete_images=int(enabled))
        await interaction.response.send_message(
            embed=ok("Server Image Filter", f"Filter gambar **{'aktif' if enabled else 'nonaktif'}** server-wide"),
            ephemeral=True,
        )

    @server_group.command(name="logchannel", description="Set channel untuk mod-log")
    @app_commands.describe(channel="Channel tujuan log")
    async def set_log_channel(self, interaction: discord.Interaction, channel: discord.TextChannel):
        await self.db.upsert_guild(interaction.guild.id, log_channel=channel.id)
        await interaction.response.send_message(
            embed=ok("Log Channel Set", f"Mod log akan dikirim ke {channel.mention}"),
            ephemeral=True,
        )

    @server_group.command(name="warn", description="Toggle DM peringatan ke user saat pesan dihapus")
    @app_commands.describe(enabled="Aktifkan atau matikan")
    async def warn_users(self, interaction: discord.Interaction, enabled: bool):
        await self.db.upsert_guild(interaction.guild.id, warn_user=int(enabled))
        await interaction.response.send_message(
            embed=ok("User Warning", f"DM peringatan **{'aktif' if enabled else 'nonaktif'}**"),
            ephemeral=True,
        )

    # ────────────────────────────────────────────────────────────────────────
    # CHANNEL GROUP
    # ────────────────────────────────────────────────────────────────────────
    channel_group = app_commands.Group(
        name="channel", description="Konfigurasi AutoMod per channel", parent=automod
    )

    @channel_group.command(name="links", description="Toggle filter link di channel tertentu")
    @app_commands.describe(channel="Target channel", enabled="Aktifkan atau matikan")
    async def channel_links(self, interaction: discord.Interaction, channel: discord.TextChannel, enabled: bool):
        await self.db.upsert_channel(channel.id, interaction.guild.id, delete_links=int(enabled))
        await interaction.response.send_message(
            embed=ok("Channel Link Filter", f"Filter link **{'aktif' if enabled else 'nonaktif'}** di {channel.mention}"),
            ephemeral=True,
        )

    @channel_group.command(name="images", description="Toggle filter gambar di channel tertentu")
    @app_commands.describe(channel="Target channel", enabled="Aktifkan atau matikan")
    async def channel_images(self, interaction: discord.Interaction, channel: discord.TextChannel, enabled: bool):
        await self.db.upsert_channel(channel.id, interaction.guild.id, delete_images=int(enabled))
        await interaction.response.send_message(
            embed=ok("Channel Image Filter", f"Filter gambar **{'aktif' if enabled else 'nonaktif'}** di {channel.mention}"),
            ephemeral=True,
        )

    @channel_group.command(name="whitelist_add", description="Izinkan domain tertentu di channel (link tidak dihapus)")
    @app_commands.describe(channel="Target channel", domain="Domain, contoh: youtube.com")
    async def whitelist_add(self, interaction: discord.Interaction, channel: discord.TextChannel, domain: str):
        domain = domain.lower().strip().removeprefix("https://").removeprefix("http://").split("/")[0]
        await self.db.add_whitelist(channel.id, interaction.guild.id, domain)
        await interaction.response.send_message(
            embed=ok("Whitelist Updated", f"`{domain}` ditambahkan ke whitelist {channel.mention}"),
            ephemeral=True,
        )

    @channel_group.command(name="whitelist_remove", description="Hapus domain dari whitelist channel")
    @app_commands.describe(channel="Target channel", domain="Domain yang ingin dihapus")
    async def whitelist_remove(self, interaction: discord.Interaction, channel: discord.TextChannel, domain: str):
        await self.db.remove_whitelist(channel.id, interaction.guild.id, domain.lower().strip())
        await interaction.response.send_message(
            embed=ok("Whitelist Updated", f"`{domain}` dihapus dari whitelist {channel.mention}"),
            ephemeral=True,
        )

    # ────────────────────────────────────────────────────────────────────────
    # ROLE GROUP
    # ────────────────────────────────────────────────────────────────────────
    role_group = app_commands.Group(
        name="role", description="Kelola role yang bypass semua filter", parent=automod
    )

    @role_group.command(name="add", description="Tambah role ke daftar exempt")
    @app_commands.describe(role="Role yang akan dibebaskan dari filter")
    async def role_add(self, interaction: discord.Interaction, role: discord.Role):
        await self.db.add_role(interaction.guild.id, role.id)
        await interaction.response.send_message(
            embed=ok("Exempt Role Added", f"{role.mention} kini bebas dari semua filter"),
            ephemeral=True,
        )

    @role_group.command(name="remove", description="Hapus role dari daftar exempt")
    @app_commands.describe(role="Role yang ingin dihapus dari exempt")
    async def role_remove(self, interaction: discord.Interaction, role: discord.Role):
        await self.db.remove_role(interaction.guild.id, role.id)
        await interaction.response.send_message(
            embed=ok("Exempt Role Removed", f"{role.mention} dihapus dari daftar exempt"),
            ephemeral=True,
        )

    @role_group.command(name="list", description="Lihat semua role yang exempt")
    async def role_list(self, interaction: discord.Interaction):
        roles = await self.db.get_roles(interaction.guild.id)
        mentions = [
            interaction.guild.get_role(rid).mention
            for rid in roles if interaction.guild.get_role(rid)
        ]
        await interaction.response.send_message(
            embed=info("Exempt Roles", ", ".join(mentions) if mentions else "Tidak ada role exempt."),
            ephemeral=True,
        )

    # ── Error Handler ─────────────────────────────────────────────────────────
    async def cog_app_command_error(self, interaction: discord.Interaction, error: app_commands.AppCommandError):
        if isinstance(error, app_commands.MissingPermissions):
            await interaction.response.send_message(
                embed=err("Permission Denied", "Kamu butuh permission **Administrator**."),
                ephemeral=True,
            )
        else:
            log.error(f"Command error: {error}", exc_info=error)
            if not interaction.response.is_done():
                await interaction.response.send_message(
                    embed=err("Error", f"`{error}`"), ephemeral=True
                )


async def setup(bot):
    await bot.add_cog(Commands(bot))
